### PR TITLE
Fix/datarace when concurrent save the same file

### PR DESCRIPTION
**Pull request by 9547**
### What problem does this PR solve? 

In some cases, saving the same file may cause an error due to a race condition, eg in the cluster deploy with TLS supported scenario,  the _ca.crt_ file was saved multi times in parallel, in my local env, this leads to the _ca.crt_ file to be left empty.

```bash
root@control:~/.tiup/storage/cluster/clusters/test_scale_core_30845# ll config-cache/ca*
-rw-r--r-- 1 root root 1123 Oct 10 13:54 config-cache/ca-2020-10-10T13:54:07.798975201Z.crt
-rw-r--r-- 1 root root 1123 Oct 10 13:54 config-cache/ca-2020-10-10T13:54:08.057441637Z.crt
-rw-r--r-- 1 root root 1123 Oct 10 13:54 config-cache/ca-2020-10-10T13:54:08.070494012Z.crt
-rw-r--r-- 1 root root 1123 Oct 10 13:54 config-cache/ca-2020-10-10T13:54:08.448460096Z.crt
-rw-r--r-- 1 root root 1123 Oct 10 13:54 config-cache/ca-2020-10-10T13:54:08.550741396Z.crt
-rw-r--r-- 1 root root 1123 Oct 10 13:54 config-cache/ca-2020-10-10T13:54:08.719875556Z.crt
-rw-r--r-- 1 root root 1123 Oct 10 13:54 config-cache/ca-2020-10-10T13:54:09.14992231Z.crt
-rw-r--r-- 1 root root 1123 Oct 10 13:54 config-cache/ca-2020-10-10T13:54:09.312662983Z.crt
-rw-r--r-- 1 root root    0 Oct 10 13:54 config-cache/ca-2020-10-10T13:54:09.315387179Z.crt
-rw-r--r-- 1 root root    0 Oct 10 13:54 config-cache/ca-2020-10-10T13:54:09.401021401Z.crt
-rw-r--r-- 1 root root 1123 Oct 10 13:54 config-cache/ca-2020-10-10T13:54:09.505369444Z.crt
-rw-r--r-- 1 root root 1123 Oct 10 13:54 config-cache/ca-2020-10-10T13:54:09.614051511Z.crt
-rw-r--r-- 1 root root    0 Oct 10 13:54 config-cache/ca-2020-10-10T13:54:09.671778643Z.crt
-rw-r--r-- 1 root root 1123 Oct 10 13:54 config-cache/ca-2020-10-10T13:54:10.837351042Z.crt
-rw-r--r-- 1 root root 1123 Oct 10 13:54 config-cache/ca-2020-10-10T13:54:10.924662339Z.crt
-rw-r--r-- 1 root root 1123 Oct 10 13:54 config-cache/ca-2020-10-10T13:54:11.240826755Z.crt
-rw-r--r-- 1 root root    0 Oct 10 13:54 config-cache/ca-2020-10-10T13:54:11.414010019Z.crt
-rw-r--r-- 1 root root 1123 Oct 10 13:54 config-cache/ca-2020-10-10T13:54:11.921334897Z.crt
-rw-r--r-- 1 root root 1123 Oct 10 13:54 config-cache/ca.crt
```

### What is changed and how it works?

introduce a simple lock to ensure the file was written in sequence.

### Check List 

Tests 

 - Unit test
 - Integration test

Code changes

 - Has exported function/method change
 - Has exported variable/fields change
 - Has interface methods change
 - Has persistent data change

Side effects

 - Possible performance regression
 - Increased code complexity
 - Breaking backward compatibility

Related changes

 - Need to cherry-pick to the release branch
 - Need to update the documentation

Release notes:

```release-note
NONE
```

### CODE REVIEW

**Comment by 9547** [comment]
@lonng PTAL

File: pkg/file/file.go [hunk 1]
```diff
@@ -25,6 +28,13 @@ func SaveFileWithBackup(path string, data []byte, backupDir string) error {
 		return errors.Errorf("%s is directory", path)
 	}
 
+	if _, ok := fileLocks[path]; !ok {
```
Comment (lucklove) [inline pkg/file/file.go:31]: The fileLocks itself is not thread safe, so I think the datarace is still exists. eg.

```bash
# assume the fileLocks is empty as initial state.
# thread 1 start
if _, ok := fileLocks["/path/a"]; !ok { # the ok is false because fileLocks is empty
# thread 1 hang after check !ok but before fileLocks[path] = &sync.Mutex{}

# thread 2 start
if _, ok := fileLocks["/path/a"]; !ok { # the ok is false because fileLocks is empty and thread 1 doesn't assign it yet
    fileLocks["/path/a"] = &sync.Mutex{} # assign a new mutex and success
}

fileLocks["/path/a"].Lock()
# thread 2 enter critical area

# thread 1 start
fileLocks["/path/a"] = &sync.Mutex{} # assign a new mutex and success
fileLocks["/path/a"].Lock() # this will success because thread 1 reset fileLocks["/path/a"]
# thread 1 enter critical area
```

In this case, both threads enter into the same critical area.

Comment (9547) [reply pkg/file/file.go:31]: Thanks for your suggestion, sorry to forget that the map needs to be locked in concurrent access.

File: Makefile [hunk 2]
```diff
@@ -10,7 +10,7 @@ GOARCH  := $(if $(GOARCH),$(GOARCH),amd64)
 GOENV   := GO111MODULE=on CGO_ENABLED=0 GOOS=$(GOOS) GOARCH=$(GOARCH)
 GO      := $(GOENV) go
 GOBUILD := $(GO) build $(BUILD_FLAG)
-GOTEST  := GO111MODULE=on CGO_ENABLED=1 $(GO) test -p 3
+GOTEST  := GO111MODULE=on CGO_ENABLED=1 go test -p 3
```
Comment (lonng) [inline Makefile:13]: It's better to revert this change.

**Review by lonng — COMMENTED** [review]
Rest LGTM

**Review by lonng — APPROVED** [review]
LGTM
